File: cuttingPolygonsPython/src/segment.py
import numpy as np
import src.point as point
import src.vertex as vertex




class PositionSide:  #funziona come enum, classe con delle variabili
    Left=0
    Right = 1
    Forward = 2
    Behind = 3
    Begin = 4
    End = 5
    In = 6


class Segment:
    TOLERANCE = 1.0E-7
    def __init__(self):
        self._from = point.Point()
        self._to = point.Point()

    def setFromPoint(self, da: point.Point, a: point.Point): #setto dal punto
        self._from = da
        self._to = a

    def setFromvertex(self, da: vertex.Vertex, a: vertex.Vertex): #setto dal vertice
        self._from = da.getPoint()
        self._to = a.getPoint()

    def getFrom(self):
        return self._from

    def getTo(self):
        return self._to

    def sidePoint(self, p: point.Point):  #point nome: tipo variabile
        position= PositionSide.Left #riga 8-9 del .cpp
        p0 = self._from
        p1 = self._to
        p2 = p
        v = point.Point()
        w = point.Point()
        k = point.Point()
        v.setcoordinate(p1.getX()-p0.getX(), p1.getY()-p0.getY())
        w.setcoordinate(p2.getX()-p0.getX(), p2.getY()-p0.getY())
        k.setcoordinate(p2.getX()-p1.getX(), p2.getY()-p1.getY())
        controllo = np.array([v.getcoordinates(), w.getcoordinates()])
        z = np.linalg.det(controllo) #determinante
        if z> self.TOLERANCE:
            position =PositionSide.Left
            return position
        elif z<- self.TOLERANCE:
            position=PositionSide.Right
            return position
        elif v.getX()*w.getX()<0 or w.getY()*v.getY()<0:  #da sistemare
            position = PositionSide.Behind
            return position
        elif v.norm2()<w.norm2():
            position = PositionSide.Forward
            return position
        elif w.norm2() >= -self.TOLERANCE and w.norm2() <= self.TOLERANCE :  #senza param
            position = PositionSide.Begin
            return position
        elif k.norm2() >= -self.TOLERANCE and k.norm2() <= self.TOLERANCE:
            position = PositionSide.End
            return position
        else:
            position = PositionSide.In
            return position

File: CuttingPolygonsC/src/segment.cpp
#include "segment.h"
namespace SegmentNamespace {



Segment::PositionSide Segment::SidePoint(const Point &point) const
{
    PositionSide position;
    position=Left;
    Point p2;
    Point p0;
    Point p1;
    p0=_from;
    p1=_to;
    p2=point;
    Point v;
    Point w;
    Point k;
    v=Point(p1.GetX()-p0.GetX(), p1.GetY()-p0.GetY());
    w=Point(p2.GetX()-p0.GetX(), p2.GetY()-p0.GetY());
    k=Point (p2.GetX()-p1.GetX(), p2.GetY()-p1.GetY()); //vedi slides geom comp
    Matrix2d controllo;
    controllo.col(0)=v.getCoordinate();
    controllo.col(1)=w.getCoordinate();
    double z;
    z= controllo.determinant();
    if (z>tolerance)
    {
        position=Left;
        return position;
    }
    else if (z<-tolerance)
    {
        position=Right;
        return position;
    }
    else if (v.GetX()*w.GetX()<0 || w.GetY()*v.GetY()<0)
    {
        position=Behind;
        return position;
    }
    else if (v.ComputeNorm2()<w.ComputeNorm2())
    {
        position =Forward;
        return position;
    }
    else if (w.ComputeNorm2()<=tolerance  && w.ComputeNorm2()>=-tolerance )
    {
        position = Begin;
        return position;
    }
    else if (k.ComputeNorm2()<=tolerance  && k.ComputeNorm2()>=-tolerance )
    {
        position=End;
        return position;
    }
    else
    {
        position =In;
        return position;
    }
}

Segment::~Segment()
{
}

Segment::Segment()
{
    _from=Point();
    _to=Point();
}

Segment::Segment(const Point &da, const Point &a)
{
    _from=da;
    _to=a;
}

Segment::Segment(const Vertex &da,const Vertex &a)
{
    _from=da.GetPoint();
    _to=a.GetPoint();
}

Point Segment::GetFrom() const
{
    return _from;
}

Point Segment::GetTo() const
{
    return _to;
}

}
